### PR TITLE
Migrate from Setuptools to Hatch to support Python 3.12

### LINKED ISSUE #1371: Update setuptools to >62
Currently pyccel uses a version of setuptools >= 37, < 61, however this means that it is common to come across the problem documented here :  https://github.com/pyccel/pyccel/discussions/1259. This was fixed in setuptools version 62, however pyccel previously had some problems moving to this version (due to pickling in the installation). We should work out how to fix these problems to avoid the documented issue

### LINKED ISSUE #1564: Python 3.12 Release
Python 3.12 was released 2 weeks ago. We should add tests for this new version

### LINKED ISSUE #1639: Cannot install Pyccel with Python 3.12
### Describe the bug
When trying to install Pyccel using Python 3.12 an error is raised:

```
Requirement already satisfied: pip in /opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/site-packages (23.3.2)
Processing /home/runner/work/pyccel/pyccel
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'error'
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      Traceback (most recent call last):
        File "/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
        File "/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 112, in get_requires_for_build_wheel
          backend = _build_backend()
                    ^^^^^^^^^^^^^^^^
        File "/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 77, in _build_backend
          obj = import_module(mod_path)
                ^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/hostedtoolcache/Python/3.12.1/x64/lib/python3.12/importlib/__init__.py", line 90, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1[31](https://github.com/pyccel/pyccel/actions/runs/7273693028/job/19818166490#step:6:32)0, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1[33](https://github.com/pyccel/pyccel/actions/runs/7273693028/job/19818166490#step:6:34)1, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 9[35](https://github.com/pyccel/pyccel/actions/runs/7273693028/job/19818166490#step:6:36), in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 994, in exec_module
        File "<frozen importlib._bootstrap>", line [48](https://github.com/pyccel/pyccel/actions/runs/7273693028/job/19818166490#step:6:49)8, in _call_with_frames_removed
        File "/tmp/pip-build-env-mrbo6mji/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 16, in <module>
          import setuptools.version
        File "/tmp/pip-build-env-mrbo6mji/overlay/lib/python3.12/site-packages/setuptools/version.py", line 1, in <module>
          import pkg_resources
        File "/tmp/pip-build-env-mrbo6mji/overlay/lib/python3.12/site-packages/pkg_resources/__init__.py", line 2178, in <module>
          register_finder(pkgutil.ImpImporter, find_on_path)
                          ^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
Error: Process completed with exit code 1.
```

The error shows that the version of `pkgutil` is incompatible with the version of `setuptools` to which we have pinned pyccel. This can be fixed by fixing #1371

### Additional context
See https://github.com/pyccel/pyccel/actions/runs/7273693028/job/19818166490

### PR DESCRIPTION
**Pull request by EmilyBourne**
Migrate from `setuptools` to `hatch` as suggested by Python's discussion board (https://discuss.python.org/t/how-to-include-pickled-files-in-installation/41527/7). Hatch is better documented and has support for intermediate build steps which create artifacts (which is required by Pyccel to pickle .pyh files).

Moving away from setuptools fixes #1371 . This allows support for Python 3.12 to be activated which fixes #1639. Tests are activated which fixes #1564

### CODE REVIEW

**Comment by EmilyBourne** [comment]
/bot run pickle editable_pickle pickle_wheel anaconda_linux anaconda_windows

File: ci_tools/bot_tools/bot_funcs.py [hunk 1]
```diff
@@ -418,22 +418,24 @@ def is_test_required(self, commit_log, name, key, state):
...
+                                                    for f in diff)
         elif key in ('spelling'):
             has_relevant_change = lambda diff: any(f.endswith('.md') or f == '.dict_custom.txt' for f in diff) #pylint: disable=unnecessary-lambda-assignment
         elif key in ('pickle', 'pickle_wheel', 'editable_pickle'):
-            has_relevant_change = lambda diff: any(f.startswith('pyccel/') and f.endswith('.py') \
-                                                    and f != 'pyccel/version.py' for f in diff) #pylint: disable=unnecessary-lambda-assignment
+            has_relevant_change = lambda diff: any((f.startswith('pyccel/') and f.endswith('.py') #pylint: disable=unnecessary-lambda-assignment
+                                                    and f != 'pyccel/version.py') or f == 'pyproject.toml'
+                                                    or f == 'MANIFEST.in' or f.startswith('install_scripts/')
+                                                    for f in diff)
```
Comment (yguclu) [inline ci_tools/bot_tools/bot_funcs.py:438]: Have you not removed `MANIFEST.in`?

File: pyproject.toml [hunk 2]
```diff
@@ -1,11 +1,56 @@
...
+pyccel-init = "pyccel.commands.pyccel_init:pyccel_init_command"
+
+[project.urls]
+Homepage = "https://github.com/pyccel/pyccel"
+
+[tool.hatch.version]
+path = "pyccel/version.py"
+
+[tool.hatch.build.targets.sdist]
+
```
Comment (yguclu) [inline pyproject.toml:48]: Why this empty section? Should we exclude some directories here?

File: ci_tools/bot_tools/bot_funcs.py (hunk 1 above)
Comment (EmilyBourne) [reply ci_tools/bot_tools/bot_funcs.py]: Fixed [b879e5c](https://github.com/pyccel/pyccel/pull/1638/commits/b879e5ce8444599d228ffb11dbc64fe71e4a47d5)

File: pyproject.toml (hunk 2 above)
Comment (EmilyBourne) [reply pyproject.toml]: It does indeed seem to be superfluous. Fixed [b0f3b5a](https://github.com/pyccel/pyccel/pull/1638/commits/b0f3b5a3a779e09300161c25b71fc2b5c2bdaaf6)

**Review by yguclu — APPROVED** [review]

**Pull request merged**